pcp/pcp.1
.\"	$Id: pcp.1,v 1.18 2007/07/03 18:33:37 garbled Exp $
.\"
.\" Copyright (c) 1998, 1999, 2000
.\"	Tim Rightnour.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"	This product includes software developed by Tim Rightnour for
.\"	The NetBSD Foundation, Inc.
.\" 4. The name of Tim Rightnour may not be used to endorse or promote 
.\"    products derived from this software without specific prior written 
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" The following requests are required for all man pages.
.Dd January 11, 2007
.Dt PCP 1
.Sh NAME
.Nm pcp
.Nd copy a file to a cluster of machines
.Sh SYNOPSIS
.Nm
.Op Fl bceprtv
.Op Fl f Ar fanout
.Op Fl g Ar rungroup1,...,rungroupN
.Op Fl o Ar porttimeout
.Op Fl n Ar portnum
.Op Fl w Ar node1,...,nodeN
.Op Fl x Ar node1,...,nodeN
.Ar source_file1
.Op source_file2 ... source_fileN
.Op destination file
.Sh DESCRIPTION
The 
.Nm
utility can be used to copy a file, or directory to a group of machines.
The group of machines can be listed on the command line, or default to
the environment setting
.Ev CLUSTER
.
The following options are available:
.Bl -tag -width www
.It Fl b
This flag causes
.Nm
to copy files in reverse, or backwards.  Instead of copying a single file
from a single machine, out to the list of target machines, the command will
copy a file from all target machines, to the local machine, and name it
file.nodename.  This can be used to gather configuration files from a group
of machines and store them locally.
.It Fl c
Normally 
.Nm 
copies files serially, to all the nodes specified.  When the
.Fl c
option is used,
.Nm
will copy the files in parallel to the nodes, subject to the
.Ev FANOUT
environment setting and
.Fl f
option.  Smaller files will benefit greatly from parallel copies,
due to reduced latency in the initial authentication, and connection.
However, large files, especially on a shared ethernet, will suffer
tremendous speed degredation, because of the collisions the transfers
will create.  This effect can be reduced, by lowering the fanout size
of the cluster.
.It Fl e
Unless the
.Fl e
option is specified, output from remote commands will not be reported 
to the user.
.It Fl p
The
.Fl p
option causes
.Nm
to attempt to preserve (duplicate) in its copies the modification
times and modes of the source files, ignoring the
.Ar umask  .
By default, the mode and owner of
.Ar file2
are preserved if it already existed; otherwise the mode of the source file
modified by the
.Xr umask  2
on the destination host is used.
.It Fl r
If any of the source files are directories,
.Nm
copies each subtree rooted at that name; in this case
the destination must be a directory.
.It Fl v
Prints the version of ClusterIt to the stdout, and exits.
.It Fl f
If the
.Fl f
option is specified, followed by a number, it sets the fanout size of 
the cluster.  The fanout size is the number of nodes a command will
run on in parallel at one time.  Thus a 80 node cluster, with a fanout 
size of 64, would run 64 nodes in parallel, then, when all have
finished, it would execute the command on the last 16 nodes.  The
fanout size defaults to 64.  This option overrides the
.Ev FANOUT
environment variable.
.It Fl o
The
.Fl o
option is used to set the timeout in seconds to be used when testing remote
connections.  The default is five seconds.
.It Fl n
The
.Fl n
option can be used to set the port number that testing should occur on when
testing remote connections.  The default behavior is to guess based on the
remote command name.
.It Fl t
The
.Fl t
option causes
.Nm
to attempt a connection test to each node prior to attempting to copy the
remote file.  If the test fails for any reason, the remote command will
not be attempted.  This can be useful when clusterfiles have suffered bitrot
and some nodes no longer exist, or might be down for maintenance.  The default
timeout is 5 seconds.  The timeout can be changed with the
.Fl o
option.
.Nm
will attempt to guess the port number of the remote service based on your
.Ev RCMD_CMD
setting.  It knows about scp and rcp.  If
.Nm
fails to guess your port correctly, you may use the
.Fl n
argument to set the remote port number.  If the
.Ev RCMD_TEST
environment variable exists, the testing will automatically take place.
.It Fl g
If the
.Fl g
option is specified, followed by a comma separated list of group
names, the command will only be run on that group of nodes.  A node
may be a part of more than one group if desired, however running
without the
.Fl g
option will run the command on the same node as many times as it
appears in the file specified by the
.Ev CLUSTER
environment variable.  This option is silently ignored if used with the
.Fl w
option.
.It Fl w
If the
.Fl w
option is specified, followed by a comma delimited list of machine names,
the
.Ar command
will be run on each node in the list.  Without this flag,
.Nm
runs on the nodes listed in the file pointed to by the
.Ev CLUSTER
environment variable.
.It Fl x
The
.Fl x
option can be used to exclude specific nodes from the cluster.  The 
format is the same as the
.Fl w
option, a comma delimited list of machine names.  This option is
silently ignored if used with the
.Fl w
option.
.El
.Sh ENVIRONMENT
.Nm
utilizes the following environment variables:
.Bl -tag -width "RCP_CMD"
.It Ev CLUSTER
Contains a filename, which is a newline separated list of nodes
in the cluster.
.It Ev RCP_CMD
Command to use to copy files to remote machines.  The command chosen must
be able to connect with no password to the remote host.  Defaults to
.Ic rcp .
.It Ev RCP_CMD_ARGS
Arguments to pass to the remote copy command. Defaults to none.
.It Ev RCP_USER
The username to connect to remote machines as by default.
.It Ev FANOUT
When set, limits the maximum number of concurrent commands sent at
once.  This can be used to keep from overloading a small host when
sending out commands in parallel.  Defaults to 64.  This environment
setting can be overridden by the
.Fl f
option.
.It Ev RCMD_PORT
The port number used to test remote connections.  See the
.Fl n
flag.
.It Ev RCMD_TEST
When set,
.Nm
will automatically test all hosts before launching the remote command. See the
.Fl t
option for more information.
.It Ev RCMD_TEST_TIMEOUT
The timeout in seconds to use when testing for remote connections.
.El
.\" .Sh FILES
.Sh EXAMPLES
The command:
.Pp
pcp file
.Pp
will copy the file 
.Sq file 
to the home directory of the current user on
all machines listed in the file pointed to by the
.Ev CLUSTER
environment setting.
.Pp
.Pp
The command:
.Pp
pcp -w hadar,rigel file
.Pp
will copy the file
.Sq file
to the home directory of the current user on the machines
.Sq hadar
and
.Sq rigel .
.Pp
.Pp
The command:
.Pp
pcp -w hadar,rigel /usr/bin/foo /usr/local/bin/foo
.Pp
will copy the file
.Sq /usr/bin/foo
to
.Sq /usr/local/bin/foo
on the machines
.Sq hadar
and
.Sq rigel .
.Pp
.Pp
The command:
.Pp
pcp -bw hadar,rigel /etc/passwd hello
.Pp
will copy the file
.Sq /etc/passwd
from both hadar and rigel, and name it
.Sq hello.hadar
and
.Sq hello.rigel
in the current working directory.
.Pp
.Pp
.Sh DIAGNOSTICS
Exit status is 0 on success, 1 if an error occurs.
.Sh SEE ALSO
.Xr cp 1 ,
.Xr dsh 1 ,
.Xr rcp 1 ,
.Xr kerberos 3 ,
.Xr hosts.equiv 5 ,
.Xr rhosts 5
.Sh HISTORY
The
.Nm
command appeared in clusterit 1.0. It is based on the
.Nm
command in IBM PSSP.
.Sh AUTHOR
.Nm Pcp
was written by Tim Rightnour.


CHANGES
$Id: CHANGES,v 1.18 2007/01/04 18:59:06 garbled Exp $

Add -s option to dsh.  This adds the ability to copy a script to a remote
machine, execute it, and delete it, all in a single operation.

Add RCP_CMD_ARGS processing to pcp.

* Release 2.4 *

Fix jsd to work properly on Linux.

Fix stdin command mode on all the utilities.

Make jsh and jsd more robust.  Keep jsd from permanently losing it's nodes.

Fix pdf to work on Linux.

Add a -v option to all programs to tell the user what version of
ClusterIt they are running.  CVS will allways have _BETA tagged on the end,
and it will be removed at release.

Add RVT_CMD_ARGS to dvt, to allow passing of args to rvt

Incorporate patches from debian xvt package into rvt:    
Fix xvt manpage typos
Add -rv (reverse-video) option 
Better colormap handling
All other patches were either irrelevant (makefile fixes, etc) or allready
incorporated.

Fix pdf to work on linux hosts

Replace most instances of sprintf with snprintf. Did not touch rvt, as I would
prefer to keep that as unmodified as possible, and most of them would not
be easy to fix anyhow.

Replace most instances of malloc with calloc.   Mostly to prevent any
unanticipated wierdness, and a few of the mallocs were immediately followed
by memset commands anyhow.

Add RCMD_CMD_ARGS to most of the programs. This makes it easier to do
things like run "ssh -4" as your RCMD_CMD.

Stop allocating the command to be run on the remote end as buf[MAXBUF],
instead compute it at runtime and dynamically allocate it to prevent
possible overflow on a huge command.

Fix a bug noted by Jot Powers, where trailing whitespace on a GROUP or LUMP
entry would cause dsh -g to not match it.

Fix a bug noted by John Bales  where linux machines often recieved truncated 
output from the child ssh/rsh process.

Fix bug where all the programs would mangle argv[0] for ps.

* Release 2.3.1 *

Fix a bug noted by James F Wilkus on OpenBSD where dsh would segv when
reading the clusterfile.

Convert seq over to rseq fully, not just for linux anymore.  Sigh.

Fix the clusterit.spec file to work properly.  Tested on redhat

* Release 2.3 *

Made all remote commands use poll instead of a while loop to read the incoming
stream of data from stdout and stderr.  This fixes bug where linux ssh will
hang up.

Fixed linux command line argument processing.

Changed the -l <username> processing to invoke RCMD_CMD with user@host rather
than -l user host.  This makes it much more reliable.

Added an option to dsh to test remote ports before connecting.  This can be
useful where a remote host might be down, and a long timeout would occur
waiting for a response.

Updated rvt to use a newer set of sources from xvt.  Fixes questionable
licensing in xvt, and makes it more reliable.

Converted makefiles to autoconf for ease of porting.  No more stupid patch
sets which were a terrible burden to maintain and generate.

ANSIfyed most of the code (other than rvt). Nobody uses KNF anymore.. sigh..

Incorporated many patches sent by users from the internet.  I have
unfortunately lost track of what patch came from who, but thier help is
greatly appreciated.  Most of these patches either fixed bugs, added stability
or fixed porting bugs.  The xvt conversion was based mainly on a patch sent
in by a user.  My thanks go out to those of you who have sent them in.

# changes since 2.2

* Release 2.2 *
Missing this data, sorry
* Release 2.0 *

Added dvt and rvt..  an interactive form of dsh with xterms.

Fixed a bug in jsd and dsh where it would lock up running ssh waiting for
input.

Add a jsd dir, and move jsd.c from dsh to the jsd dir.  Add nearly complete
jsh program, and jcommon.[ch].

Add new common subdir and move a tonnage of crap in there.  Mainly
common.c, common.h and strsep.c.

Fix horrid bug in dsh where dsh | more would lock terminals up.  Still
need to fix in seq, run, probably pcp.

Add <netinet/if_ether.h> to barrierd.c to shut AIX C up.

Lots of minor bugfixes, mostly dealing with failures to open files, opening
0 byte clusters, etc etc..

Pcp was still using it's own code to parse strings and groups.  Fully rewrote
the back-end internals to use the new linked-list stuff, and utilize common.c.
Pcp's code is now virtually identical to dsh, whereas before it had diverged
significantly.

Error handling has finally graduated to where I no longer feel that it
"needs work".

Fix a place where malloced data was strcatted to without being bzeroed.

Group, exclusion and -w nodes are now dynamicly generated arrays.  Removes
the MAX_GROUPS restrictions.

The nodelist and grouplists are no longer maintained in any way similar to
the original setup.  Nodes are not stored in a flat array of nodes, with a
corresponding flat array of group names.  Nodes are now stores in a linked
list of machines to run on.  Nodes which are not to be run on, are not
included in this list, so the back-end routines only need to execute code
on every node they are handed in the linked list.  This is all handled in
common.c now.  This removes the all compile time restrictions on MAXNODES,
allowing the program to scale much more efficiently.

# changes since 1.3
* Release 1.3 *

Break dsh, run and seq's common code into a common.c and common.h file,
so I don't forget to add options to run and seq when fixing dsh.  Also, fix
some locations where run and seq were out of sync slightly.

Fix a bug in pdf where if the -m option was used, filesystems at 100%
would not be displayed.

Handle ^C's correctly.  If a remote process is hung, and it is the current
process being read, you may ^C it, and the interrupt will be sent to the
remote process.  This would be just like hitting ^C on the remote end, 
and will only affect the process you are currently reading from.

Added ability to pcp multiple files and wildcards to remote hosts.

Made pcp and dsh compile on an ANSI compiler (IBM C, SUN C)

Make output from dsh look nicer, and format well.

# changes since 1.2
* Release 1.2 *


pcp/pcp.c
/* $Id: pcp.c,v 1.28 2007/07/03 18:33:37 garbled Exp $ */
/*
 * Copyright (c) 1998, 1999, 2000
 *	Tim Rightnour.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Tim Rightnour.
 * 4. The name of Tim Rightnour may not be used to endorse or promote 
 *    products derived from this software without specific prior written 
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <fcntl.h>
#include <poll.h>
#include <libgen.h>
#include <signal.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/resource.h>
#include <sys/wait.h>
#include "../common/common.h"

#if !defined(lint) && defined(__NetBSD__)
__COPYRIGHT(
"@(#) Copyright (c) 1998, 1999, 2000\n\
        Tim Rightnour.  All rights reserved.\n");
__RCSID("$Id: pcp.c,v 1.28 2007/07/03 18:33:37 garbled Exp $");
#endif

extern int errno;

void do_copy(char **argv, int recurse, int preserve, char *username);
void paralell_copy(char *rcp, int nrof, char *username, char *source_file,
    char *destination_file);
void serial_copy(char *rcp, char *username, char *source_file,
    char *destination_file);
void reverse_copy(char *rcp, char *username, char *source_file,
    char *destination_file);

char **lumplist;
char **rungroup;
char *progname;
int fanout, concurrent, quiet, debug, grouping, exclusion, nrofrungroups;
int testflag, rshport, porttimeout, bflag;
node_t *nodelink;
group_t *grouplist;
volatile sig_atomic_t alarmtime;

/* 
 *  pcp is a cluster management tool based on the IBM tool of the
 *  same name.  It allows a user, or system administrator to copy files
 *  to a cluster of machines with a single command.
 */

int
main(int argc, char **argv)
{
    extern char *optarg;
    extern int optind;
    extern char *version;

    int someflag, ch, i, preserve, recurse;
    char *p, *nodename, *username, *group;
    char **exclude;
    node_t *nodeptr;

    fanout = 0;
    quiet = 1;
    concurrent = 0;
    someflag = 0;
    preserve = 0;
    recurse = 0;
    exclusion = 0;
    grouping = 0;
    nrofrungroups = 0;
    testflag = 0;
    rshport = 0;
    bflag = 0;
    porttimeout = 5; /* 5 seconds to port timeout */
    username = NULL;
    group = NULL;
    nodeptr = NULL;
    nodelink = NULL;
    exclude = NULL;

    rungroup = calloc(GROUP_MALLOC, sizeof(char **));
    if (rungroup == NULL)
	bailout();

    progname = strdup(basename(argv[0]));

#if defined(__linux__)
    while ((ch = getopt(argc, argv, "+?bcdeprtvf:g:l:n:o:w:x:")) != -1)
#else
    while ((ch = getopt(argc, argv, "?bcdeprtvf:g:l:n:o:w:x:")) != -1)
#endif
	switch (ch) {
	case 'b':               /* set reverse/backwards mode */
	   bflag = 1;
	   break;
	case 'c':		/* set concurrent mode */
	    concurrent = 1;
	    break;
	case 'd':		/* hidden debug mode */
	    debug = 1;
	    break;
	case 'e':		/* display error messages */
	    quiet = 0;
	    break;
	case 'p':		/* preserve file modes */
	    preserve = 1;
	    break;
	case 'r':		/* recursive directory operations */
	    recurse = 1;
	    break;
	case 't':           /* test the nodes before connecting */
	    testflag = 1;
	    break;
	case 'n':           /* what is the rsh port number? */
	    rshport = atoi(optarg);
	    break;
	case 'o':               /* set the test timeout in seconds */
	    porttimeout = atoi(optarg);
	    break;
	case 'l':               /* invoke me as some other user */
	    username = strdup(optarg);
	    break;
	case 'f':		/* set the fanout size */
	    fanout = atoi(optarg);
	    break;
	case 'g':		/* pick a group to run on */
	    grouping = 1;
	    nrofrungroups = parse_gopt(optarg);
	    break;
	case 'x':		/* exclude nodes, w overrides this */
	    exclusion = 1;
	    exclude = parse_xopt(optarg);
	    break;
	case 'w':		/* perform operation on these nodes */
	    someflag = 1;
	    i = 0;
	    for (p = optarg; p != NULL; ) {
		nodename = (char *)strsep(&p, ",");
		if (nodename != NULL)
		    (void)nodealloc(nodename);
	    }
	    break;
	case 'v':
	    (void)printf("%s: %s\n", progname, version);
	    exit(EXIT_SUCCESS);
	    break;
	case '?':
	    (void)fprintf(stderr,
		"usage: %s [-ceprv] [-f fanout] [-g rungroup1,...,rungroupN] "
		"[-l username] [-x node1,...,nodeN] [-w node1,..,nodeN] "
		"source_file1 [source_file2 ... source_fileN] "
		"[desitination_file]\n", progname);
	    return(EXIT_FAILURE);
	    /* NOTREACHED */
	    break;
	default:
	    break;
	}
    if (fanout == 0 && getenv("FANOUT"))
	    fanout = atoi(getenv("FANOUT"));
    else
	    fanout = DEFAULT_FANOUT;
    if (username == NULL && getenv("RCP_USER"))
	    username = strdup(getenv("RCP_USER"));
    if (!rshport && getenv("RCP_PORT"))
	    rshport = atoi(getenv("RCP_PORT"));
    if (!testflag && getenv("RCMD_TEST"))
	    testflag = 1;
    if (porttimeout == 5 && getenv("RCMD_TEST_TIMEOUT"))
	    porttimeout = atoi(getenv("RCMD_TEST_TIMEOUT"));

    rshport = get_rshport(testflag, rshport, "RCP_CMD");
    
    if (!someflag)
	parse_cluster(exclude);
    
    argc -= optind;
    argv += optind;
    do_copy(argv, recurse, preserve, username);
    return(EXIT_SUCCESS);
}

/* 
 * Do the actual dirty work of the program, now that the arguments
 * have all been parsed out.
 */

void do_copy(char **argv, int recurse, int preserve, char *username)
{
    int numsource, j, nrofargs;
    size_t len, rem;
    char *source_file, *destination_file, *tempstore, **rcp, *rcpstring;
    char **argvp;
    node_t *nodeptr;

    if (debug) {
	j = 0;
	if (username != NULL)
	    (void)printf("As User: %s\n", username);
	(void)printf("On nodes:\n");
	for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next) {
	    if (!(j % 4) && j > 0)
		(void)printf("\n");
	    (void)printf("%s\t", nodeptr->name);
	    j++;
	}
    }
    if (*argv == (char *)NULL) {
	(void)fprintf(stderr, "Must specify at least one file to copy\n");
	exit(EXIT_FAILURE);
    }
    argvp = argv;
    len = strlen(*argvp) + 2;
    while (*++argvp != NULL) {
	    len += 1; /* space */
	    len += strlen(*argvp);
    }

    source_file = calloc(len, sizeof(char));
    rem = len;

    tempstore = NULL;
    strncpy(source_file, *argv, rem);
    rem -= strlen(*argv);
    numsource = 1;
    while (*++argv != NULL) {
	numsource++;
	if (tempstore != NULL) {
	    (void)strncat(source_file, " ", rem);
	    rem -= 1;
	    (void)strncat(source_file, tempstore, rem);
	    rem -= strlen(tempstore);
	}
	tempstore = *argv;
    }
    if (numsource == 1)
	destination_file = strdup(source_file);
    else
	destination_file = strdup(tempstore);

    if (bflag && numsource > 2) {
	fprintf(stderr, "Can only specify one file for reverse copy\n");
	bailout();
    }

    if (bflag && numsource == 1) {
	    free(destination_file);
	    destination_file = strdup(".");
    }
    
    if (debug)
	printf("\nDo Copy: %s %s\n", source_file, destination_file);

    rcp = parse_rcmd("RCP_CMD", "RCP_CMD_ARGS", &nrofargs);
    j = nrofargs;
    rcpstring = build_rshstring(rcp, nrofargs);
    if (recurse) {
	    strcat(rcpstring, " -r ");
	    nrofargs++;
    }
    if (preserve) {
	    strcat(rcpstring, " -p ");
	    nrofargs++;
    }

    if (bflag)
	    reverse_copy(rcpstring, username, source_file, destination_file);
    else if (concurrent)
	    paralell_copy(rcpstring, nrofargs + numsource, username,
		source_file, destination_file);
    else
	    serial_copy(rcpstring, username, source_file, destination_file);
    free(source_file);
    free(destination_file);
    for (nrofargs=0; nrofargs < j; nrofargs++)
	    free(rcp[nrofargs]);
    free(rcp);
}

/* Copy files in paralell.  This is preferred with smaller files, because
   the initial connection and authentication latency is longer than the
   file transfer.  With large files, you generate more collisions than
   good packets, and actually slow it down, thus serial is faster. */

void
paralell_copy(char *rcp, int nrof, char *username, char *source_file,
	      char *destination_file)
{
    int i, j, n, g, status, pollret, fdf;
    char *rcpstring, pipebuf[2048], *cd;
    FILE *fd, *fda, *in;
    char **argz, **aps;
    node_t *nodeptr, *nodehold;
    size_t maxnodelen, rcplen;
    pid_t currentchild;
    struct pollfd fds[2];

    j = i = maxnodelen = 0;
    in = NULL;
    cd = pipebuf;

    argz = calloc(nrof + 3, sizeof(char *));
    if (argz == NULL)
	    bailout();
    
    for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next) {
	if (strlen(nodeptr->name) > maxnodelen)
	    maxnodelen = strlen(nodeptr->name);
	j++;
    }

    if (username)
	    rcplen = strlen(username) + 128;
    else
	    rcplen = 128;
    rcplen += strlen(source_file) + strlen(rcp) + maxnodelen +
	strlen(destination_file);
    rcpstring = calloc(rcplen, sizeof(char));
    if (rcpstring == NULL)
	    bailout();
    
    i = j;
    j = i / fanout;
    if (i % fanout)
	j++;

    g = 0;
    nodeptr = nodelink;
    for (n=0; n <= j; n++) {
	nodehold = nodeptr;
	for (i=0; (i < fanout && nodeptr != NULL); i++) {
	    g++;
	    /*
	     * we set up pipes for each node, to prepare for the oncoming
	     * barrage of data. Close on exec must be set here, otherwise
	     * children spawned after other children, inherit the open file
	     * descriptors, and cause the pipes to remain  open forever.
	     */
	    if (pipe(nodeptr->out.fds) != 0)
		bailout();
	    if (pipe(nodeptr->err.fds) != 0)
		bailout();
	    if (fcntl(nodeptr->out.fds[0], F_SETFD, 1) == -1)
		bailout();
	    if (fcntl(nodeptr->out.fds[1], F_SETFD, 1) == -1)
		bailout();
	    if (fcntl(nodeptr->err.fds[0], F_SETFD, 1) == -1)
		bailout();
	    if (fcntl(nodeptr->err.fds[1], F_SETFD, 1) == -1)
		bailout();
	    if (username != NULL)
		(void)snprintf(rcpstring, rcplen, "%s %s %s@%s:%s", rcp,
			       source_file, username, nodeptr->name,
			       destination_file);
	    else
		(void)snprintf(rcpstring, rcplen, "%s %s %s:%s", rcp,
			       source_file, nodeptr->name, destination_file);
	    if (debug)
		printf("Running command: %s\n", rcpstring);
	    nodeptr->childpid = fork();
	    switch (nodeptr->childpid) {
	    case -1:
		bailout();
		break;
	    case 0:
		if (dup2(nodeptr->out.fds[1], STDOUT_FILENO) != STDOUT_FILENO)
		    bailout();
		if (dup2(nodeptr->err.fds[1], STDERR_FILENO) != STDERR_FILENO)
		    bailout();
		if (close(nodeptr->out.fds[0]) != 0)
		    bailout();
		if (close(nodeptr->err.fds[0]) != 0)
		    bailout();
		/* stdin & stderr non-blocking */
		fdf = fcntl(nodeptr->out.fds[0], F_GETFL);
		fcntl(nodeptr->out.fds[0], F_SETFL, fdf|O_NONBLOCK);
		fdf = fcntl(nodeptr->err.fds[0], F_GETFL);
		fcntl(nodeptr->err.fds[0], F_SETFL, fdf|O_NONBLOCK);
		
		if (testflag && rshport > 0 && porttimeout > 0)
			if (!test_node_connection(rshport, porttimeout,
						  nodeptr))
			    exit(EXIT_SUCCESS);
		for (aps = argz; (*aps = strsep(&rcpstring, " ")) != NULL;)
		    if (**aps != '\0')
			++aps;
		execvp(argz[0], argz);
		bailout();
	    default:
		break;
	    } /* switch */
	    nodeptr = nodeptr->next;
	} /* for i */
	nodeptr = nodehold;
	for (i=0; (i < fanout && nodeptr != NULL); i++) {
	    currentchild = nodeptr->childpid;
	    /* now close off the useless stuff, and read the goodies */
	    if (close(nodeptr->out.fds[1]) != 0)
		bailout();
	    if (close(nodeptr->err.fds[1]) != 0)
		bailout();
	    fda = fdopen(nodeptr->out.fds[0], "r"); /* stdout */
	    if (fda == NULL)
		bailout();
	    fd = fdopen(nodeptr->err.fds[0], "r"); /* stderr */
	    if (fd == NULL)
		bailout();
	    fds[0].fd = nodeptr->out.fds[0];
	    fds[1].fd = nodeptr->err.fds[0];
	    fds[0].events = POLLIN|POLLPRI;
	    fds[1].events = POLLIN|POLLPRI;
	    pollret = 1;

	    while (pollret >= 0) {
		int gotdata;
		
		pollret = poll(fds, 2, 5);
		gotdata = 0;
		if ((fds[0].revents&POLLIN) == POLLIN ||
		    (fds[0].revents&POLLHUP) == POLLHUP ||
		    (fds[0].revents&POLLPRI) == POLLPRI) {
		    while ((cd = fgets(pipebuf, sizeof(pipebuf), fda))) {
			if (!quiet)
			    (void)printf("%*s: %s", -maxnodelen,
				nodeptr->name, cd);
			gotdata++;
		    }
		}
		if ((fds[1].revents&POLLIN) == POLLIN ||
		    (fds[1].revents&POLLHUP) == POLLHUP ||
		    (fds[1].revents&POLLPRI) == POLLPRI) {
		    while ((cd = fgets(pipebuf, sizeof(pipebuf), fd))) {
			if (!quiet)
				(void)printf("%*s: %s", -maxnodelen,
				    nodeptr->name, cd);
			gotdata++;
		    }
		}
		if (!gotdata)
		    if (((fds[0].revents&POLLHUP) == POLLHUP ||
			 (fds[0].revents&POLLERR) == POLLERR ||
			 (fds[0].revents&POLLNVAL) == POLLNVAL) &&
			((fds[1].revents&POLLHUP) == POLLHUP ||
			 (fds[1].revents&POLLERR) == POLLERR ||
			 (fds[1].revents&POLLNVAL) == POLLNVAL))
			break;
	    }
	    fclose(fda);
	    fclose(fd);
	    (void)wait(&status);
	    nodeptr = nodeptr->next;
	} /* pipe read */
    } /* for n */
    free(argz);
    free(rcpstring);
}

/* serial copy */

void
serial_copy(char *rcp, char *username, char *source_file,
	    char *destination_file)
{
	node_t *nodeptr;
	char *command;
	size_t maxnodelen, rcplen;

	maxnodelen = 0;

	for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next) {
		if (strlen(nodeptr->name) > maxnodelen)
			maxnodelen = strlen(nodeptr->name);
	}
	if (username)
		rcplen = strlen(username) + 128;
	else
		rcplen = 128;
	rcplen += strlen(source_file) + strlen(rcp) + maxnodelen +
	    strlen(destination_file);
	command = calloc(rcplen, sizeof(char));
	if (command == NULL)
		bailout();
    
	for (nodeptr=nodelink; nodeptr != NULL; nodeptr = nodeptr->next) {
		if (testflag && rshport > 0 && porttimeout > 0)
			if (!test_node_connection(rshport, porttimeout,
				nodeptr))
			    continue;
		if (username != NULL)
			(void)snprintf(command, rcplen, "%s %s %s@%s:%s", rcp,
			    source_file, username, nodeptr->name,
			    destination_file);
		else
			(void)snprintf(command, rcplen, "%s %s %s:%s", rcp,
			    source_file, nodeptr->name, destination_file);
		if (debug)
			printf("Running command: %s\n", command);
		system(command);
	}
	free(command);
}

/* Reverse copy */

void
reverse_copy(char *rcp, char *username, char *source_file,
	    char *destination_file)
{
	node_t *nodeptr;
	char *command, *bname;
	struct stat st;
	size_t maxnodelen, rcplen;
	int isdir=0;

	maxnodelen = 0;

	for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next) {
		if (strlen(nodeptr->name) > maxnodelen)
			maxnodelen = strlen(nodeptr->name);
	}

	if (stat(destination_file, &st) == 0)
		if (st.st_mode && S_IFDIR)
			isdir = 1;

	bname = basename(source_file);
	
	if (username)
		rcplen = strlen(username) + 128;
	else
		rcplen = 128;
	rcplen += strlen(source_file) + strlen(rcp) + maxnodelen +
	    maxnodelen + strlen(destination_file);
	command = calloc(rcplen, sizeof(char));
	if (command == NULL)
		bailout();
    
	for (nodeptr=nodelink; nodeptr != NULL; nodeptr = nodeptr->next) {
		if (testflag && rshport > 0 && porttimeout > 0)
			if (!test_node_connection(rshport, porttimeout,
				nodeptr))
			    continue;
		if (isdir && username != NULL)
			(void)snprintf(command, rcplen, "%s %s@%s:%s %s/%s.%s",
			    rcp, username, nodeptr->name, source_file,
			    destination_file, bname, nodeptr->name);
		else if (!isdir && username != NULL)
			(void)snprintf(command, rcplen, "%s %s@%s:%s %s.%s",
			    rcp, username, nodeptr->name, source_file,
			    destination_file, nodeptr->name);
		else if (isdir)
			(void)snprintf(command, rcplen, "%s %s:%s %s/%s.%s",
			    rcp, nodeptr->name, source_file, destination_file,
			    bname, nodeptr->name);
		else
			(void)snprintf(command, rcplen, "%s %s:%s %s.%s", rcp,
			    nodeptr->name, source_file, destination_file,
			    nodeptr->name);
		if (debug)
			printf("Running command: %s\n", command);
		system(command);
	}
	free(command);
}
